### ISSUE TITLE
The Commit Transaction Workflows task can be used to circumvent validation

### ISSUE THREAD

**Issue by Piedone** [OrchardCMS/OrchardCore#6413]
The Commit Transaction task of Workflows will, well, commit the DB transaction immediately. This can be used to commit the transaction even in a state where normally it wouldn't be committed. Depending on what kind of Workflows events are or will be available this can be used to put the application into a corrupted state. This includes at least the ability to circumvent user input validation for content items but opens up the possibility of any other state corruption with security implications as well.

Simple demonstration:

![image](https://user-images.githubusercontent.com/1976647/84554407-f2f1b000-ad17-11ea-81a9-76527e2c9365.png)

Now when you save a Blog Post with a validation error the changes will still be persisted to the database. Unless those users who can edit workflows can't be completely trusted (both to not be malicious and not to make mistakes) this seems very risky. This can be a problem in any setup but will be one in every SaaS one.

My proposal would be to at least put the Commit Transaction task (and any risky future ones) into its separate feature (or module) so they can be disabled completely.

Otherwise, I'm not sure I see where this task would be useful. Where would you utilize such a low-level operation when creating a workflow? Even as a developer you have to think twice when committing the ambient transaction explicitly so it seems to me that surfacing this on the admin UI is dangerous.

**Verdict:** So as we discussed:

- For now, we can put the Commit Transaction task into its own feature so its usage can be easily prevented if you don't need it.
- https://github.com/OrchardCMS/OrchardCore/issues/6455

**Comment by jtkech** [comment]
It has been introduced through #4530 to fix #4255 where a workflow creates an item, then triggers another WF through an http activity, then waits an http response from this nested WF that also creates an item. Here even the 2 WFs are running in different scopes, they are waiting for each other because of the logic of the 2 WFs. Here a part of one of my comment in the related issue

> Here the parent worflow creates an item in a given scope and, before this scope ends and commit the session, it does an HttpRequestTask that triggers a nested workflow that also create an item (here it is just when it ends and save its finished workflow instance). The problem is that the nested workflow, to be able to save its finished instance, it is waiting for the parent scope to finish, but the parent workflow is waiting for the nested workflow to finish to continue the execution flow, and so on.

> With SQLite we can create items in 2 different scopes and, when it is done at the same time, a given scope 2 wait for the other scope 1 to finish. But if for another reason the scope 1 also wait for the scope 2 to finish, there is a dead lock. Here the other reason is a blocking activity.

---

There are other potential issues with WF events / tasks, e.g. when a WF content updated event is followed by a WF content update task that calls the updated handler that triggers again the WF content updated event. That's why in #6054, to prevent these kind of scenarios, we check if a content task is executed inline from a content event and from which event so that we can decide what we can do.

**So here we could add the same kind of checkings before committing the transaction.**

Note: In #6054 i limited my changes to content events / tasks to not have a too big PR.

**Comment by Piedone** [comment]
Thank you for your detailed reply.

I've read through https://github.com/OrchardCMS/OrchardCore/issues/4255 and my idea instead of a Commit Transaction task would also be to run all tasks (perhaps all activities) in their own transaction, i.e. own scope, as you suggested.

However, I'm not sure I understand the initial issue: So there the WF running in scope 1 issues a request to the same Orchard instance and runs a workflow there, creating (as it is with every request) scope 2. However, regardless of what's in the two WFs this shouldn't cause a deadlock. Why is it doing that though? A new content item having been created but its transaction not yet committed shouldn't prevent another scope (i.e. transaction/request) to do the same. If it does then it's an issue for all concurrent transactions, not just this very specific use-case.

**Comment by jtkech** [comment]
E.g. sqlite doesn't support concurrent writes so, if in a scope 1 there is a write in progress, a concurrent write in a scope 2 will have to wait that scope 1 commits its transaction, but that's okay if scope 1 doesn't wait for something else that relies on scope 2. The problem was that WF2 was waiting for WF1 to commit **but WF1 was also waiting for WF2 to complete its flow** => dead lock.

> run all tasks (perhaps all activities) in their own transaction, i.e. own scope, as you suggested.

Yes, that would be a solution unless if we still allow a content activity to update a content item that has been mutated by drivers, **this in the same scope / transaction**, e.g. a `ContentCreatedEvent` (triggered by a content handler) followed by a `UpdateContentTask`.

In that case the update content task (acting as a driver) is executed inline from a content handler, so it should not call content manager methods because they would call handlers that would trigger again WF content events and so on. The content manager call handlers but handlers are not intended to call content manager methods. That's why in #6054 i had to check this.

**Comment by Piedone** [comment]
Oh OK, so this is an SQLite limitation. It seems to me that what we need for SQLite is [`BEGIN CONCURRENT`](https://www.sqlite.org/cgi/src/doc/begin-concurrent/doc/begin_concurrent.md) then. However, this is [not released yet](https://stackoverflow.com/a/54475279/220230).

But I see #6054 now, thanks for explaining. Will the Commit Transaction task still be necessary after #6054?

**Comment by jtkech** [comment]
> Will the Commit Transaction task still be necessary after #6054?

Yes because of the issue we talked about, but a least we could add some checkings in the commit transaction task (as done in #6054) to see if it is executed inline from a content handler (without any intermediate waiting activity) meaning that a content item is being mutated in the current scope. But we would need to do the same with e.g. user events and any other future event meaning that some entity is being mutated.

**When not executed inline from such an handler**, seems to be okay

- This is the WF itself that retrieves the entity and then if a WF task update this entity it has to do its standalone validation and if it fails to do a session cancel so that a subsequent commit transaction task would have no effect. As it is done in content tasks with the new `ValidateAsync()` that calls validate handlers to still do some important validations when drivers are not involved (also e.g. through API controllers), and that does a session cancel if the validation fails.

**When executed inline from such an handler**, only a suggestion

- Maybe all `Xxxed` handler events would need to be called only if there is no model error (which is not the case) so that, through the related triggered WF events, a WF would only be executed if the model is valid, and as said above, if a WF task does its own mutation it would have to do its standalone validation with a possible session cancel before executing an eventual commit transaction task. Note: Just saw that user handlers are called by user drivers ! Some model checkings are done before calling these handlers but not all, other checkings are done after, maybe they would also have to be done before.

**Comment by Piedone** [comment]
Hmm, I feel like this needs a lot of plumbing to work around issues that bring more issues and the whole thing gets really complex and requires maintenance with every related future change. Could we talk about this on the Thursday meeting?

**Comment by jtkech** [comment]
Okay, not the best time for me but i'll try to be free for the next triage meeting

Hmm, can you remind me at which hour?

Note: The last part of my above comment was just a suggestion, maybe i talk too much ;)

**Comment by Piedone** [comment]
It's at 5 PM UTC (so two hours earlier than the Tuesday meeting).

**Comment by Piedone** [comment]
And it'll be here this time, new room: https://teams.microsoft.com/l/meetup-join/19%3ameeting_MjgxNGRiMDgtMWUzYy00YzRjLTllYWYtODRhMDJlNWU5YTM5%40thread.v2/0?context=%7b%22Tid%22%3a%2272f988bf-86f1-41af-91ab-2d7cd011db47%22%2c%22Oid%22%3a%22cff2928b-da0f-4c65-9cc8-742e3c4de50e%22%7d

**Comment by Piedone** [comment]
It's now @jtkech if you can join :).

**Comment by Piedone** [comment]
Thanks for coming @jtkech!

So as we discussed:

- For now, we can put the Commit Transaction task into its own feature so its usage can be easily prevented if you don't need it.
- https://github.com/OrchardCMS/OrchardCore/issues/6455

**Comment by jtkech** [comment]
@Piedone Just for info

> Now when you save a Blog Post with a validation error the changes will still be persisted to the database.

No, if the session has been canceled nothing will be persisted when committing it, and now, even if the transaction is committed and a new one is created in the context of the same scoped session instance, we now keep the `_cancel` state of a given session instance for its whole lifetime.

Hmm yes, maybe still a problem if the WF activity `CommitTransactionTask` is executed after a session save has been done, and before another validation process. But when a part of code does a session save, normally just before (maybe sometimes right after) a validation should have already been executed, and the session canceled if the validation failed.

Hmm yes, if you have a loop on multiple items doing a session save on each iteration, triggering in the middle some WFs including the commit task, and you don't want to persist anything at the end of the loop if any item validation failed.

So, i agree with the suggestion, i will do it.

> For now, we can put the Commit Transaction task into its own feature so its usage can be easily prevented if you don't need it.

**Comment by Piedone** [comment]
At the time I tested whether the invalid state is indeed persisted to the DB and it was. If it is not, now, then great!

**Comment by jtkech** [comment]
@Piedone 

Just retried the simple WF in your first comment

I edited a blog post and changed the slug to make a duplicate so that the validation will fail on publishing

In that case the WF is not executed, so that's okay

But when i re-edit it, the already created draft version still have the duplicate slug, but not related to the WF that was not executed. Then from the admin actions i discarded the draft and re-edit it, and i saw again the original correct slug

That said this is not the correct behavior, we should not persist the draft if not valid, or at least not be able to publish it from the admin actions (that's okay when publishing from the edit page that redo a validation on publishing).

This is not related to the WF, when i discard the WF it is the same, so i will open a separate issue.

**Comment by jtkech** [comment]
**Update**

@Piedone 

Yes, you were right, the updated event is triggered regardless the validation result, i didn't see it because the completed WF instance was not persisted because of the cancel done by the validation, but the draft was persisted because of the commit activity (at this point a session save on the WF instance is not yet done).

And when i really remove the commit task, that's okay

So we should not trigger the Updated Event when the model is not valid

I will update the related PR

**Comment by Piedone** [comment]
Thank you!
